File: CompetitiveProgramming/Hackerrank/Algorithms/Medium/emas_supercomputer.cpp
#include <iostream>
#include <algorithm>
#include <set>
#include <string>
#include <vector>
#include <hackerrank.hpp>

class Point
{
public:
    int x;
    int y;

    Point(int x_val, int y_val)
    {
        x = x_val;
        y = y_val;
    }

    bool operator< (const Point &right) const
    {
        return x < right.x && y < right.y;
    }

    std::string to_str() const
    {
        return "(" + std::to_string(x) + "," + std::to_string(y) + ")";
    }
};


class PlusCross
{
public:
    int area;
    std::set<Point> tiles;

    PlusCross(int area_val)
    {
        area = area_val;
    }

    void add_tiles(std::set<Point> tile_coords)
    {
        tiles.insert(tile_coords.begin(), tile_coords.end());
    }

    bool overlaps_with(PlusCross other)
    {
        for (std::set<Point>::iterator it = tiles.begin(); it != tiles.end(); it++)
        {
            if (other.tiles.find(*it) != other.tiles.end())
                return true;
        }

        return false;
    }

    void print_plus_cross()
    {
        std::cout << "AREA: " << area << std::endl;
        std::cout << "TILES: " << std::endl;

        for (auto it = tiles.begin(); it != tiles.end(); it++)
        {
            std::cout << (*it).to_str() << std::endl;
        }
    }
};


int two_pluses(std::vector<std::string>);
std::vector<PlusCross> find_crosses_from_source(std::vector<std::string>,
                                                int,
                                                int,
                                                int,
                                                int);

bool are_part_of_cross(std::vector<std::string>, int, int, int);
bool are_in_grid(int, int, int, int, int);

int main(void)
{
    std::string first_multiple_input_temp;
    std::getline(std::cin, first_multiple_input_temp);

    std::vector<std::string> first_multiple_input =
        split(rtrim(first_multiple_input_temp));

    int n = std::stoi(first_multiple_input[0]);
    int m = std::stoi(first_multiple_input[1]);

    std::vector<std::string> grid(n);

    for (int i = 0; i < n; i++) {
        std::string grid_item;
        std::getline(std::cin, grid_item);
        grid[i] = grid_item;
    }

    int result = two_pluses(grid);
    std::cout << result << std::endl;
    return 0;
}

int two_pluses(std::vector<std::string> grid)
{
    int highest_prod_area = 0;
    std::vector<PlusCross> pluses;

    // First, we need to find all the possible pluses/crosses in the grid we
    // we were given.

    for (int i = 0; i < grid.size(); i++)
    {
        std::string row = grid[i];

        for (int j = 0; j < row.length(); j++)
        {
            // Not a "good" tile, so nothing to search here.
            if (row[j] != 'G')
                continue;

            std::vector<PlusCross> tile_crosses =
                find_crosses_from_source(grid, i, j, grid.size(), row.length());
            pluses.insert(pluses.end(), tile_crosses.begin(), tile_crosses.end());
        }
    }

    // Now, we find the two pluses/crosses whose area product is the maximum one.

    for (int i = 0; i < pluses.size(); i++)
    {
        for (int j = i + 1; j < pluses.size(); j++)
        {
            if (pluses[i].overlaps_with(pluses[j]))
                continue;

            highest_prod_area = std::max(highest_prod_area,
                                         pluses[i].area * pluses[j].area);
        }
    }

    return highest_prod_area;
}

std::vector<PlusCross> find_crosses_from_source(
    std::vector<std::string> grid,
    int x,
    int y,
    int rows,
    int columns)
{
    // We have to keep track of all coordinates taken by the current plus/cross,
    // as there will be overlaps. Like for example, a 5x5 cross also includes
    // a 3x3 and a 1x1 one by definition.

    std::set<Point> tiles_used;
    std::vector<PlusCross> pluses;
    int total_area = 1;
    int i = 1;

    // The current point counts as a 1x1 plus/cross, so we add it to the set of
    // tiles visited, and the pluses/crosses we found starting from here.

    tiles_used.emplace(Point(x, y));
    PlusCross orig_cross(total_area);
    orig_cross.add_tiles(tiles_used);

    pluses.push_back(orig_cross);

    // In this loop, we grow each edge of the plus/cross by one and test if it
    // still is a valid plus/cross. If yes, then we record it into a new object.
    // If not, then we end the loop and we're done with this tile as an origin.

    while (are_in_grid(x, y, i, rows, columns) && are_part_of_cross(grid, x, y, i))
    {
        tiles_used.emplace(Point(x + i, y));
        tiles_used.emplace(Point(x - i, y));
        tiles_used.emplace(Point(x, y + i));
        tiles_used.emplace(Point(x, y - i));

        total_area += 4;
        i++;

        PlusCross next_cross(total_area);
        next_cross.add_tiles(tiles_used);
        pluses.push_back(next_cross);
    }

    return pluses;
}

bool are_part_of_cross(std::vector<std::string> grid, int x, int y, int i)
{
    return (grid[x+i][y] == 'G'
            && grid[x-i][y] == 'G'
            && grid[x][y+i] == 'G'
            && grid[x][y-i] == 'G');
}

bool are_in_grid(int x, int y, int i, int rows, int columns)
{
    return (x + i < rows && x - i >= 0 && y + i < columns && y - i >= 0);
}

File: Hackerrank/Medium/emas_supercomputer.py
#!/usr/bin/env python3

from typing import Self

class PlusCross:
    area: int
    tiles: set[tuple[int, int]]

    def __init__(self: Self, area_val: int, tiles_set: set[tuple[int, int]]):
        self.area = area_val
        self.tiles = tiles_set

    def overlaps_with(self: Self, other: Self) -> bool:
        for pt in other.tiles:
            if pt in self.tiles:
                return True
        return False


def is_shape_in_grid(x: int, y: int, delta: int, num_rows: int, num_cols) -> bool:
    return x + delta < num_rows \
        and x - delta >= 0 \
        and y + delta < num_cols \
        and y - delta >= 0


def is_a_cross(grid: list[str], x: int, y: int, delta: int) -> bool:
    return grid[x+delta][y] == 'G' \
        and grid[x-delta][y] == 'G' \
        and grid[x][y+delta] == 'G' \
        and grid[x][y-delta] == 'G'


def find_pluses_crosses_from_src(grid: list[str], row: int, col: int) -> list[PlusCross]:
    tile_coords = set()
    pluses_crosses = []
    total_area = 1
    delta = 1

    # Adding it here because we need the tuple to be a single element and the
    # constructor takes it as two.
    tile_coords.add((row, col))
    pluses_crosses.append(PlusCross(total_area, tile_coords.copy()))

    # Next, we grow each side of the shape by one. If it is still a plus/cross
    # within the grid, then we record it as another option. Otherwise, we're
    # done with this tile as a center/origin.

    while is_shape_in_grid(row, col, delta, len(grid), len(grid[row])) \
          and is_a_cross(grid, row, col, delta):
        tile_coords.add((row + delta, col))
        tile_coords.add((row - delta, col))
        tile_coords.add((row, col + delta))
        tile_coords.add((row, col - delta))

        total_area += 4
        delta += 1
        pluses_crosses.append(PlusCross(total_area, tile_coords.copy()))

    return pluses_crosses


def two_pluses(grid: list[str]) -> int:
    product_area = 0
    plus_crosses = []

    # First, we have to find all the pluses/crosses in the grid we were given.

    for i in range(len(grid)):
        for j in range(len(grid[i])):
            if grid[i][j] != 'G':
                continue
            plus_crosses += (find_pluses_crosses_from_src(grid, i, j))

    # Next, we have to find the two largest pluses/crosses that do not overlap
    # with each other, and the product of their areas is our result.

    for i in range(len(plus_crosses)):
        for j in range(i+1, len(plus_crosses)):
            if plus_crosses[i].overlaps_with(plus_crosses[j]):
                continue
            product_area = max(product_area,
                               plus_crosses[i].area * plus_crosses[j].area)

    return product_area


def main():
    first_multiple_input = input().rstrip().split()
    n = int(first_multiple_input[0])
    m = int(first_multiple_input[1])

    grid = []

    for _ in range(n):
        grid_item = input()
        grid.append(grid_item)

    result = two_pluses(grid)
    print(str(result))


if __name__ == '__main__':
    main()
